#184581177 Authorization: protected routes (all) should only be accessed by autheticated users

1MuhireDavid opened the pull request:
#### What does this PR do?
This PR adds authorization to protected routes so that they can be accessed by authorized uses.

#### Description of Task to be completed?
- Add a middleware to authorize protected routes according to their role. If your authorized you willl be able to continue else you will get a message access denied

#### How should this be manually tested?
- Clone this repo git clone https://github.com/atlp-rwanda/e-comm-team-amigos-bn
- Switch your working directory cd e-comm-team-amigos-bn
- Create an .env file and update it with .env-example
- Install all the dependencies by running npm install
- Then run the project by typing npm start

#### Endpoints to be tested:
- `user/check`

#### Any background context you want to provide?
- Have Node Js installed in your local machine

#### What are the relevant pivotal tracker/Trello stories?

[#184581177](https://www.pivotaltracker.com/n/projects/2632829/stories/184581177)

#### Screenshots (if appropriate)
 image for test
<img width="807" alt="Screen Shot 2023-03-17 at 13 40 54" src="https://user-images.githubusercontent.com/39764771/225894880-9e40e521-2762-4d1a-b833-77e69b1938e0.png">
<img width="1440" alt="Screen Shot 2023-03-17 at 13 48 29" src="https://user-images.githubusercontent.com/39764771/225896547-59a28182-dc45-40f2-8761-61ae72d36e1d.png">
<img width="1440" alt="Screen Shot 2023-03-17 at 13 48 39" src="https://user-images.githubusercontent.com/39764771/225896587-f44c550e-aa58-44f9-83c6-8a8056c8d696.png">
<img width="1440" alt="Screen Shot 2023-03-17 at 13 48 20" src="https://user-images.githubusercontent.com/39764771/225896592-62c6a0db-c8d1-4466-9735-230d0c34947a.png">
<img width="1435" alt="Screen Shot 2023-03-17 at 13 47 39" src="https://user-images.githubusercontent.com/39764771/225896602-191b9578-1f2c-4df0-a300-fd6118b3fb3c.png">

GasanaJr committed:
feat(user): User Password Update

- add a password update endpoint

[Finishes #184581183]

ch(authorization): Add authorization middleware

[Finishes #184581177]

ch(user): add authorization

- add authorize middleware to be used on protected routes

[Finishes #18451177]

ch(authorization): Add authorization middleware

[Finishes #184581177]

nshutijonathan reviewed (commented):
Please narrow down your commit messages to two

Emile-SHUMBUSHO commented on src/middleware/authValidation.js:
```diff
@@ -99,9 +101,26 @@ const resetPassValidator = async (req, res, next) => {
     res.status(400).send(error.message);
   }
 };
+const authorize = (roles) => asyncHandler(async (req, res, next) => {
+  const authHeader = await req.get('Authorization');
+  if (!authHeader) {
+    return res.status(401).json({ error: 'No token provided!' });
+  }
+
+  const token = authHeader.split(' ')[1];
+  const decodedToken = jwt.verify(token, process.env.SECRET_KEY);
+  const role = decodedToken.role;
+  if (!roles.includes(role)) {
+    return res.status(401).json({
+      error: 'Access denied! You are not allowed to perform this operation.'
+    });
+  }
+  next();
+});
```
authorize function has been already declared from `middleware/verifyToken` use them as a middleware it's purpose is to avoid writing a lot of codes with the same functionality.

GasanaJr replied:
On line 105, change `const authHeader = await req.get('Authorization')` to `const authHeader = await req.get('auth-token')`. This is because swagger doesn't allow to name a header Authorization and remember you need to embed the token in the headers

Emile-SHUMBUSHO commented on src/routes/auth/index.js:
```diff
@@ -11,5 +10,10 @@ router.get('/verify_email/:token', authentication.emailVerification);
 router.post('/otp', authentication.checkotp);
-router.patch("/updatePassword",verifyToken, authentication.updatePassword)
+router.patch('/updatePassword', verifyToken, authentication.updatePassword);
+router.get('/check', validate.authorize(['admin', 'vendor', 'normal']), (req, res) => {
+  res.status(200).json({
+    message: 'You are authorized',
+  });
+});
```
if validate.authorize works good it's fine but if we have a lot of condition to check from here the routes directory will become more complex as project keeps growing better to define all these conditions from middle directory and use it here as a middleware function. that's my suggestion.

AbdulKhaliq59 reviewed (commented):
Resolve CircleCI issues

Emile-SHUMBUSHO commented:
This PR need to rebase from develop and a fix for failing test

1MuhireDavid committed:
feat(user): add authorization

- create authorize middleware
- add authorize middleware to a protected route like update password

[Delivers #184581177]

feat(user): add authorization

- create authorize middleware
- add authorize middleware to a protected route like update password

[Delivers #184581177]

GasanaJr reviewed (dismissed):
Fix the circleci tests

nshutijonathan reviewed (approved):
LGTM

ericndungutse reviewed (approved):
lGTM

GasanaJr committed:
feat(user): Users Public Chat Space

- add chat feature to the e-commerce website
- create and endpoint to get all messages sent by users

[Finishes #184581204]

Habinezajanvier reviewed (approved):
👍
